Objects/typeobject.c

```diff
@@ -3066,11 +3150,16 @@ tail_contains(PyObject *tuple, Py_ssize_t whence, PyObject *o)
 static PyObject *
 class_name(PyObject *cls)
 {
+#ifdef Py_GIL_DISABLED
+    // Avoid re-entrancy and use tp_name
+    return type_name(cls, NULL);
+#else
```

Comment:
This seems fishy

Reply:
Maybe a little.  :-)   The return value is only used for the type name in the error message about having duplicate classes in the MRO.  I think using `tp_name` for that error is okay.

Keeping the logic that effectively does `return cls.__name__` is more complicated.   We would have to start-the-world, if stopped and then re-stop it again after the getattr call.